### ISSUE TITLE
Volumes may never be attached when detaching failed temporarily 

### ISSUE THREAD

**Issue by llinvokerl** [kubernetes/kubernetes#111000]
### What happened?

Recently, the cpu load of my **kube-apiserver** gets higher after I expanded more pods and nodes into the cluster.

And I found that some volumes declared by pods could not be mounted after the restart of pods, even though the pods were successful scheduled. here's the kubectl describe information:

```
$kubectl describe pod chaos-bing-test-5-c0-2
Events:
  Type     Reason       Age    From                        Message
  ----     ------       ----   ----                        -------
  Normal   Scheduled    3m55s  default-scheduler           Successfully assigned default/chaos-bing-test-5-c0-2 to i22d01261.eu95sqa
  Warning  FailedMount  112s   kubelet, i22d01261.eu95sqa  Unable to mount volumes for pod "chaos-bing-test-5-c0-2_default(2680bd56-fd3a-11ec-8d16-70e28411919b)": timeout expired waiting for volumes to attach or mount for pod "default"/"chaos-bing-test-5-c0-2". list of unmounted volumes=[chaos-bing-test-5-c0-pvc]. list of unattached volumes=[chaos-bing-test-5-c0-pvc shm default-token-5js9b delta-config control-conf]
```

After checking the logs of **attach_detach_controller**, I found that in `reconcile.go`,`nodeStatusUpdater.UpdateNodeStatuses()` return an error, thus the `attacherDetacher.DetachVolume()` was skipped. That should be related to the high load on my **kube-apiserver** as it occasionally returns 500 http code (but this may not be the root cause). Here is the error message:

```
{"log":"I0705 12:10:07.634904       1 node_status_updater.go:86] Could not update node status for \"a13m05557.na61\"; re-marking for update. failed to patch status \"{\\\"status\\\":{\\\"volumesAttached\\\":null}}\" for node \"a13m05557.na61\": request declared a Content-Length of 35 but only wrote 0 bytes\n","stream":"stderr","time":"2022-07-05T12:10:07.634930166Z"}
{"log":"E0705 12:10:07.634925       1 reconciler.go:219] UpdateNodeStatuses failed while attempting to report volume as attached for volume \"pvc-751b518c-aa59-11ec-8772-7cd30ac02cb8\" (UniqueName: \"kubernetes.io/csi/csi-hostpath^79ff946c-aa59-11ec-8926-7cd30ab71afc\") on node \"a13m05557.na61\" : failed to patch status \"{\\\"status\\\":{\\\"volumesAttached\\\":null}}\" for node \"a13m05557.na61\": request declared a Content-Length of 35 but only wrote 0 bytes\n","stream":"stderr","time":"2022-07-05T12:10:07.634948215Z"}
```

**Here comes the problem:** 
In my opinion, the next loop of `reconcile()` would retry `nodeStatusUpdater.UpdateNodeStatuses()` and `attacherDetacher.DetachVolume()`, since the state of **actualStateOfWorld** is different from the state of **desiredStateOfWorld**. but it didn't happened.

By recording some additional logs, I notice that the pod was scheduled to the destination node by the scheduler right after above error message appeared, which cause `attach_detach_controller.go:podUpdate()` executed. That would write the volume info back to `desiredStateOfWorld.nodesManaged[NodeName].volumesToAttach[VolumeName]`. After that, the next loop of `reconcile()` would skip the detach operation for that volume, since **desiredStateOfWorld** was 'equal to' **actualStateOfWorld**. The log snippet is as follows:

```
...
I0706 14:44:19.597962       1 actual_state_of_world.go:384] Set detach request time to current time for volume kubernetes.io/csi/csi-hostpath^50aaebd1-b0ca-11ec-8c38-70e2840fcf1f on node "i22d01261.eu95sqa"
... 
// waiting unmount
...
E0706 14:44:31.992432       1 reconciler.go:218] UpdateNodeStatuses failed while attempting to report volume as attached for volume "pvc-500613f4-b0ca-11ec-9c0d-70e28411919b" (UniqueName: "kubernetes.io/csi/csi-hostpath^50aaebd1-b0ca-11ec-8c38-70e2840fcf1f") on node "i22d01261.eu95sqa" : failed to patch status \"{\\\"status\\\":{\\\"volumesAttached\\\":null}}\" for node \"i22d01261.eu95sqa\": request declared a Content-Length of 35 but only wrote 0 bytes\n"...
...
// my extra log shows the pod was scheduled at this time.
I0706 14:44:34.332395 call attach_detach_controller.go:podUpdate() and write desiredStateOfWorld.nodesManaged[nodeName].volumesToAttach[volumeName]
...
I0706 14:44:34.373217       1 reconciler.go:265] Volume attached--touching for volume "pvc-500613f4-b0ca-11ec-9c0d-70e28411919b" (UniqueName: "kubernetes.io/csi/csi-hostpath^50aaebd1-b0ca-11ec-8c38-70e2840fcf1f") from node "i22d01261.eu95sqa"
I0706 14:44:34.373217       1 reconciler.go:265] Volume attached--touching for volume "pvc-500613f4-b0ca-11ec-9c0d-70e28411919b" (UniqueName: "kubernetes.io/csi/csi-hostpath^50aaebd1-b0ca-11ec-8c38-70e2840fcf1f") from node "i22d01261.eu95sqa"
I0706 14:44:34.474256       1 reconciler.go:265] Volume attached--touching for volume "pvc-500613f4-b0ca-11ec-9c0d-70e28411919b" (UniqueName: "kubernetes.io/csi/csi-hostpath^50aaebd1-b0ca-11ec-8c38-70e2840fcf1f") from node "i22d01261.eu95sqa"
...
```

Finally, **attach_detach_controller** would not attach the volume neither, because the states of **desiredStateOfWorld** and **actualStateOfWorld** are the same. **attach_detach_controller** mistakenly thought the volume was attached, but it' not the case.

### What did you expect to happen?

when `nodeStatusUpdater.UpdateNodeStatuses()` fail, **attach_detach_controller** should retry to detach the volume until the the detach operation completed. However, it was interrupted by the update of **desiredStateOfWorld**.

### How can we reproduce it (as minimally and precisely as possible)?

`kubernetes/pkg/controller/volume/attachdetach/reconciler/reconciler.go:`
``` kubernetes/pkg/controller/volume/attachdetach/reconciler/reconciler.go
			klog.Errorf("force skip detaching")
			continue
			// Update Node Status to indicate volume is no longer safe to mount.
			err = rc.nodeStatusUpdater.UpdateNodeStatuses()
			if err != nil {
				// Skip detaching this volume if unable to update node status
				klog.Errorf(attachedVolume.GenerateErrorDetailed("UpdateNodeStatuses failed while attempting to report volume as attached", err).Error())
				continue
			}
```
1. Insert a log and `continue` before `rc.nodeStatusUpdater.UpdateNodeStatuses()`
2. Restart  a pod  mounting pv

Then you'll found log message "force skip detaching" would disappear after a few loops of `reconcile()`

### Kubernetes version

<details>

```console
$ kubectl version
Client Version: version.Info{Major:"1", Minor:"13", GitVersion:"v1.13.3", GitCommit:"721bfa751924da8d1680787490c54b9179b1fed0", GitTreeState:"clean", BuildDate:"2019-02-01T20:08:12Z", GoVersion:"go1.11.5", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"13", GitVersion:"v1.13.3", GitCommit:"721bfa751924da8d1680787490c54b9179b1fed0", GitTreeState:"archive", BuildDate:"2021-05-10T07:53:22Z", GoVersion:"go1.12.17", Compiler:"gc", Platform:"linux/amd64"}
```

</details>

### Cloud provider

<details>
N/A
</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

**Comment by 249043822** [comment]
If `nodeStatusUpdater.UpdateNodeStatuses()` fails, that means the volume has not been detached and the volumeattachment is still on the nodeStatus, then kubelet should see that the volume still attaches to the node, so I'm not understand why your pod failed to mount, it may not be caused the controller.
what's the error for kubelet?

**Comment by llinvokerl** [comment]
> If `nodeStatusUpdater.UpdateNodeStatuses()` fails, that means the volume has not been detached and the volumeattachment is still on the nodeStatus, then kubelet should see that the volume still attaches to the node, so I don't understand why your pod failed to mount, it may not be caused by the controller. what's the error for kubelet?

![20220718114609](https://user-images.githubusercontent.com/38915183/179443032-236f3154-1933-466e-9e8b-82df6dd18a10.jpg)

There are `two nodeStatusUpdater.UpdateNodeStatuses()`. The one before `DetachVolume()` was failed, but another which is at the end of `reconciler()` was successfully executed. Thus cause the kubelet coundn't get the volume in `node.Status`.

**Comment by llinvokerl** [comment]
> If `nodeStatusUpdater.UpdateNodeStatuses()` fails, that means the volume has not been detached and the volumeattachment is still on the nodeStatus, then kubelet should see that the volume still attaches to the node, so I don't understand why your pod failed to mount, it may not be caused by the controller. what's the error for kubelet?

kubelet log repeat the message below infinitely:
```
reconciler.go:207] operationExecutor.VerifyControllerAttachedVolume started for volume...
nestedpendingoperations.go:267] Operation for xxx failed. No retries permitted until xxx. Error: "Volume not attached according to node status for volume xxx pod xxx"
```

**Comment by 249043822** [comment]
Okay, I know what you said, actually this problem is similar with the one which was fixed by https://github.com/kubernetes/kubernetes/pull/104526, but it is not completely.
Give another pr below to fix this.

**Comment by arvindbd** [comment]
@llinvokerl @249043822 @jingxu97 Can we have the fix (https://github.com/kubernetes/kubernetes/pull/111205) back-ported to 1.21? 

Given that this completes the https://github.com/kubernetes/kubernetes/pull/104526, the initial fix for which was [backported to 1.21](https://github.com/kubernetes/kubernetes/pull/106053), it would make sense to have this follow-up fix also be added to 1.21. Looks like there is a [1.21.15-rc0](https://github.com/kubernetes/kubernetes/tree/v1.21.15-rc.0) that we could get this into. 

We have a 100+ clusters deployed in production that are running 1.21. We see this bug at least once a week and it's a headache to deal with. Upgrading to 1.22 or higher is a big undertaking for us.
